Userland/watch.cpp

```diff
@@ -0,0 +1,171 @@
+#include <LibCore/ArgsParser.h>
+#include <stdio.h>
+#include <sys/time.h>
+#include <sys/wait.h>
+#include <time.h>
+#include <unistd.h>
+
+static int opt_interval = 2;
+static bool flag_noheader = false;
+static bool flag_beep_on_fail = false;
+static int exit_code = 0;
+static pid_t child_pid = -1;
+
+void handle_signal(int signal);
+int run_command(const Vector<const char*>& command);
+int64_t usecs_till(const struct timeval& now, const struct timeval& targ);
+struct timeval get_current_time();
+String build_header_string(const Vector<const char*>&);
+
+int main(int argc, char** argv)
+{
+    Vector<const char*> command;
+    Core::ArgsParser args_parser;
+    args_parser.add_option(opt_interval, "Amount of time between updates", "interval", 'n', "seconds");
+    args_parser.add_option(flag_noheader, "Turn off the header describing the command and interval", "no-title", 't');
+    args_parser.add_option(flag_beep_on_fail, "Beep if the command has a non-zero exit code", "beep", 'b');
+    args_parser.add_positional_argument(command, "Command to run", "command");
+    args_parser.parse(argc, argv);
+    auto header = build_header_string(command);
+    command.append(nullptr);
+
+    signal(SIGINT, handle_signal);
+
+    struct timeval interval {
+        opt_interval, 0
+    };
+    struct timeval now = get_current_time();
+    struct timeval next_run_time = now;
+    bool ran_immediately = false;
+    while (true) {
+        int usecs_to_sleep = usecs_till(now, next_run_time);
+        ran_immediately = usecs_to_sleep < 0;
+        while (usecs_to_sleep > 0) {
+            usleep(usecs_to_sleep);
+            now = get_current_time();
+            usecs_to_sleep = usecs_till(now, next_run_time);
+        }
+        // clear screen, reset cursor
+        printf("\033[H\033[2J");
+        // print header
+        if (!flag_noheader) {
+            printf("%s\n\n", header.characters());
+        }
+        fflush(stdout);
+        if (run_command(command) != 0) {
+            exit_code = 1;
+            if (flag_beep_on_fail) {
+                printf("\a");
+                fflush(stdout);
+            }
+        }
+        if (ran_immediately) {
+            // calculate next run time based on current time
+            AK::timeval_add(now, interval, next_run_time);
+        } else {
+            // calculate next run time based on previous run time
+            AK::timeval_add(next_run_time, interval, next_run_time);
+        }
+        now = get_current_time();
+    }
+}
+
+String build_header_string(const Vector<const char*>& command)
+{
+    StringBuilder builder;
+    builder.appendf("Every %d.0s: \'", opt_interval);
+    for (auto i = 0u; i < command.size(); ++i) {
+        if (i > 0) {
+            builder.append(' ');
+        }
+        builder.append(command[i]);
+    }
+    builder.append("\'");
+    return builder.build();
+}
+
+struct timeval get_current_time()
+{
+    struct timespec ts {
+    };
+    struct timeval tv {
+    };
```

Comment:
Here and in `usecs_till`: Just drop the `{}`? imho it would look better.

Reply:
Fixed